plotly/figure_factory/_bullet.py

```diff
@@ -0,0 +1,257 @@
+            fig['data'].append(bar)
+
+        # measures bars
+        for idx in range(len(df.iloc[row]['measures'])):
+            inter_colors = colors.n_colors(
+                measure_colors[0], measure_colors[1],
+                len(df.iloc[row]['measures']), 'rgb'
+            )
+            x = ([sorted(df.iloc[row]['measures'])[-1 - idx]] if as_rows
+                 else [0.5])
+            y = ([0.5] if as_rows
+                 else [sorted(df.iloc[row]['measures'])[-1 - idx]])
+            bar = go.Bar(
+                x=x,
+                y=y,
+                marker=dict(
+                    color=inter_colors[-1 - idx]
+                ),
+                name='measures',
+                hoverinfo='x' if as_rows else 'y',
+                orientation='h' if as_rows else 'v',
+                width=0.4,
+                base=0,
+                xaxis='x{}'.format(row + 1),
+                yaxis='y{}'.format(row + 1)
+            )
+            fig['data'].append(bar)
+
+        # markers
+        x = df.iloc[row]['markers'] if as_rows else [0.5]
+        y = [0.5] if as_rows else df.iloc[row]['markers']
+        markers = go.Scatter(
+            x=x,
+            y=y,
+            marker=dict(
+                color='rgb(0, 0, 0)',
+                symbol=marker_symbol,
+                size=marker_size
+            ),
+            name='markers',
+            hoverinfo='x' if as_rows else 'y',
+            xaxis='x{}'.format(row + 1),
+            yaxis='y{}'.format(row + 1)
+        )
+        fig['data'].append(markers)
+
+        # labels
+        title = df.iloc[row]['title']
+        if 'subtitle' in df:
+            subtitle = '<br>{}'.format(df.iloc[row]['subtitle'])
+        else:
+            subtitle = ''
+        label = '<b>{}</b>'.format(title) + subtitle
+        annot = utils.annotation_dict_for_label(
+            label,
+            (num_of_lanes - row if as_rows else row + 1),
+            num_of_lanes, subplot_spacing,
+            'row' if as_rows else 'col',
+            True if as_rows else False,
+            False
+        )
+        fig['layout']['annotations'].append(annot)
+
+    return fig
+
+
+def create_bullet(df, as_rows=True, marker_size=16,
+                  marker_symbol='diamond-tall', range_colors=None,
+                  measure_colors=None, subplot_spacing=None,
+                  title='Bullet Chart', height=600, width=1000):
+    """
+    Returns figure for bullet chart.
+
+    :param (pd.DataFrame | list) df: either a JSON list of dicts or a pandas
+        DataFrame. All keys must be one of 'title', 'subtitle', 'ranges',
+        'measures', and 'markers'.
+    :param (bool) as_rows: if True, the bars are placed horizontally as rows.
+        If False, the bars are placed vertically in the chart.
+    :param (int) marker_size: sets the size of the markers in the chart.
+    :param (str | int) marker_symbol: the symbol of the markers in the chart.
+        Default='diamond-tall'
```

Comment:
similarly here, can we make this more generic with something like `scatter_options=None` that get recursively merged into the trace:

```
fig = create_bullet(df, scatter_options={'marker': {'size': 12, 'symbol': 'diamond-tall'}, 'name': 'Bullet'})
```

By making it more abstract, we allow the user to pass in whatever they want and we don't have to keep adding new attributes when they are requestd (e.g. what if the user wants a line around their markers, or wants to change the name of the trace, or wants to change the color of the marker - instead of having `marker_line_color=None, name=None, marker_color=None`, we could just let the user pass in an object that we recursively merge onto the appropriate trace)

Reply:
by recursively merge i mean deep merge